### ISSUE TITLE
Show comments directly on the diff

### ISSUE THREAD

**Issue by amit1rrr** [ReviewNB/support#34]
We show all posted comments on the discussion tab alongside the cell for which the comment was written.  It could be useful to also show the comments on the notebook diff itself. That way users can see more of the notebook context (than just the single cell). 

Please note that NOT all comments can be shown on the diff since new commits coming into the PR can merge/split/delete the cell itself. GitHub follows the same rule, it shows whatever comments it can on the diff and the rest are shown under conversations tab marked as _Outdated_

E.g. I wrote 3 review comments for [this PR](https://github.com/ReviewNB/treon/pull/6) and only one is visible on the diff.

[More discussion here](https://github.com/ReviewNB/support/issues/31 )

***
_Feel free to upvote/downvote the issue indicating whether you think this is useful feature or not. I also welcome additional questions/comments/discussion on the issue._

**Comment by sid-kap** [comment]
+1, are you still thinking about implementing this? I personally think approach (1) that you mentioned in #31 sounds more useful:
> Match GitHub's approach and show the comments where we can. This has the risk of not having context on most comments.

**Comment by amit1rrr** [comment]
Thanks Sid. It's still on the radar but a low priority at the moment. Part of the reason is the fact that even if we implement it, it'll only show a small percentage of comments on the diff (comments where the context file hasn't changed after the comment was posted).

Is not having comments directly on the diff causing problem in your review workflow? Or is it more like enhancement to how you review today?

**Comment by sid-kap** [comment]
Thanks for your response! I think it would be more of an enhancement, not a major problem at the moment.

When you say "the context file hasn't changed", do you mean that if the notebook file has changed at all, even in a different cell, then it's no longer possible to associate the old comment with the cell?

**Comment by amit1rrr** [comment]
> do you mean that if the notebook file has changed at all, even in a different cell, then it's no longer possible to associate the old comment with the cell?

That is right. It's not impossible but it's hard because we do not store diffs (or any repository contents for that matter). We just say 7th cell in this notebook has comment X and that information (cell_index=7) is also stored in your GitHub comment itself (as a non-visible div tag). Let me explain with an example.

Let's say you wrote 2 comments on my notebook, one near the top & another near the bottom cell. Now I push some changes to address top comment. In the process I add some new cells & delete/merge some cells. Now even though the bottom cell isn't modified it's cell_index in the notebook has likely changed. So now to associate the old comment with new bottom cell we have to compare notebook contents pre & post commit & find the cell where source lines matches exactly. But it's tricky because there can be 2 different cells with exact same source lines (e.g. fig.show() or just print(x)). 

To summarize, it's not impossible to map comments made in earlier commits to the latest version of notebook BUT 1) it's tricky, so we might not be able to map every comment 2) the additional processing will slow down the API of our already slow'ish app (thanks to complete reliance on GitHub APIs & our promise to not store/cache any repository data)

**Comment by amit1rrr** [comment]
We've released this feature now where review comments are shown on notebook diffs (in addition to Discussions tab). It closely follows GitHub's review model in that we show conversations on notebook diff where we've _decent_ confidence that this was the cell the comment was originally meant for (even though the cell is now slightly modified to address review comments).

Additionally, there are hyperlinks on each conversation to jump from Discussions tab to same conversation on the notebook diff itself. This would help you see additional notebook context, see if the author has addressed your comments in a satisfactory manner. These URL's are also handy to share the conversation over internal chat channels & involve more team members in the discussion.

Hope this improves your notebook review experience. Please report any issues you see with this new feature or otherwise. Thank you for you patience.

cc: @sid-kap @juhoautio 
(Self hosted instances should get this update in their next release)

**Comment by sid-kap** [comment]
Amazing!! Really looking forward to trying it out

**Comment by sid-kap** [comment]
@amit1rrr Do you know if Affirm has the version with this feature? We recently did an upgrade, which seems to have some new features (like the folder navigator on the homepage), but I'm not seeing any comments show up on the "Changes" tab of a pull request

**Comment by amit1rrr** [comment]
No, we haven't released this feature for on-prem customers as yet. We plan to release this early next week. I will keep you posted.

**Comment by sid-kap** [comment]
Thanks for the update!

**Comment by amit1rrr** [comment]
@sid-kap New on-prem release (v0.2.17) which contains this feature is now out. Your IT group is notified of the update.
